<launch>
    
    <!-- SYSTEM MODEL -->
    <?ignore
    <param name="m"             type="double"   value="2.424" />
    <param name="rho"             type="double"   value="1.225" />    
    <param name="lr"            type="double"   value="0.1203" />
    <param name="lf"            type="double"   value="0.1377" />
    <param name="Cm0"            type="double"   value="9.46853" />
    <param name="Cm1"            type="double"   value="0.667237" />
    <param name="C0"            type="double"   value="2.61049" />
    <param name="C1"            type="double"   value="-0.00213596" />
    <param name="Cd_A"            type="double"   value="-0.466602" />
    <param name="Caf"            type="double"   value="1.3958" />
    <param name="Car"            type="double"   value="1.6775" />
    <param name="Iz"            type="double"   value="0.02" />
    ?>




    <param name="m"             type="double"   value="2.483" />
    <param name="rho"             type="double"   value="1.225" />    
    <param name="lr"            type="double"   value="0.1203" />
    <param name="lf"            type="double"   value="0.1377" />
    <param name="Cm0"            type="double"   value="10.1305" />
    <param name="Cm1"            type="double"   value="1.05294" />
    <param name="C0"            type="double"   value="3.68918" />
    <param name="C1"            type="double"   value="0.0306803" />
    <param name="Cd_A"            type="double"   value="-0.657645" />
    <param name="Caf"            type="double"   value="1.3958" />
    <param name="Car"            type="double"   value="1.6775" />
    <param name="Iz"            type="double"   value="0.04" />

    <?ignore
    <param name="Caf"            type="double"   value="40.62927783" />
    <param name="Car"            type="double"   value="69.55846999" />
    <param name="Iz"            type="double"   value="1.01950479" />
    ?>


        <!-- SYSTEM MODEL -->
    <?ignore
    <param name="m"             type="double"   value="2.424" />
    <param name="rho"             type="double"   value="1.225" />    
    <param name="lr"            type="double"   value="0.1203" />
    <param name="lf"            type="double"   value="0.132" />
    <param name="Cm0"            type="double"   value="10.1305" />
    <param name="Cm1"            type="double"   value="1.05294" />
    <param name="C0"            type="double"   value="3.68918" />
    <param name="C1"            type="double"   value="0.0306803" />
    <param name="Cd_A"            type="double"   value="-0.657645" />
    <param name="Caf"            type="double"   value="1.3958" />
    <param name="Car"            type="double"   value="1.6775" />
    <param name="Iz"            type="double"   value="0.04" />
    ?>

    <param name="max_vel"       type="double" value="5.0" />
    <param name="min_vel"       type="double" value="-5.0" />
    <!-- should be > 1.0 -->
    <param name="init_vx"     type="double" value="0.0"/>
    <param name="halfWidth"     type="double"   value="0.4" />
    <param name="trackShape"  type="str"    value="oval_iri"/>
    <param name="duty_th"       type="double"   value="0.05" />  <!-- deadzone where the wheel doesn't rotate -->


    <arg name="DISTURBANCE" default="0"/>

    <arg name="NOISE" default="1"/>

    <!-- SIMULATOR -->
    <node pkg="simulator" type="vehicleSimulator.py" name="simulator" output="screen">
    <!--node pkg="simulator" type="vehicleSimulator_comparison.py" name="simulator" output="screen"-->


        <param name="publish_frequency" type="double" value="100.0" />
        
        <param name="visualization"    type="bool"      value="False" />

        <!-- sensor freq update -->
        <!-- Generally 1 per second or 1 Hz not more than 10Hz in case of advanced gps made for UAV/AGV?? -->         
        <param name="gps_freq_update" type="double" value="100.0" />
        <param name="fcam_freq_update" type="double" value="40" />
        <param name="imu_freq_update" type="double" value="100.0" />
        <param name="enc_freq_update" type="double" value="100.0" />
        
        <!-- delay in the simulator -->
        <!-- why delay in simulator is needed?? -->
        <param name="lowLevelDyn" type="bool" value="false" />
        <param name="delay_df"    type="double" value="0.0" />
        <param name="delay_a"     type="double" value="0.0" />

        <!-- simulator disturbance parameters -->
        <param name="disturbance_on" type="bool" value="1" if="$(arg DISTURBANCE)"/>
        <param name="n_bound_pr"     type="double" value="2.0" if="$(arg DISTURBANCE)"/>
        <param name="x_std_pr"       type="double" value="0.02" if="$(arg DISTURBANCE)"/>
        <param name="y_std_pr"       type="double" value="0.02" if="$(arg DISTURBANCE)"/>
        <param name="psi_std_pr"     type="double" value="0.02" if="$(arg DISTURBANCE)"/>
        <param name="vx_std_pr"       type="double" value="0.1" if="$(arg DISTURBANCE)"/>
        <param name="vy_std_pr"       type="double" value="0.045" if="$(arg DISTURBANCE)"/>
        <param name="psiDot_std_pr"  type="double" value="0.035" if="$(arg DISTURBANCE)"/> -->

        <param name="disturbance_on" type="bool" value="0" unless="$(arg DISTURBANCE)"/>
        <param name="n_bound_pr"     type="double" value="0.0" unless="$(arg DISTURBANCE)"/>
        <param name="x_std_pr"       type="double" value="0.0" unless="$(arg DISTURBANCE)"/>
        <param name="y_std_pr"       type="double" value="0.0" unless="$(arg DISTURBANCE)"/>
        <param name="psi_std_pr"     type="double" value="0.0" unless="$(arg DISTURBANCE)"/>
        <param name="vx_std_pr"       type="double" value="0.0" unless="$(arg DISTURBANCE)"/>
        <param name="vy_std_pr"       type="double" value="0.0" unless="$(arg DISTURBANCE)"/>
        <param name="psiDot_std_pr"  type="double" value="0.0" unless="$(arg DISTURBANCE)"/> -->


        <!-- simulator sensor noise -->        
        <param name="n_bound"     type="double" value="1.0" if="$(arg NOISE)"/>
        <param name="x_std"       type="double" value="0.05" if="$(arg NOISE)"/>
        <param name="y_std"       type="double" value="0.05" if="$(arg NOISE)"/>
        <param name="yaw_std"     type="double" value="0.01" if="$(arg NOISE)"/>
        <param name="vx_std"       type="double" value="0.1" if="$(arg NOISE)"/>
        <param name="vy_std"       type="double" value="0.08" if="$(arg NOISE)"/>
        <param name="omega_std"  type="double" value="0.1" if="$(arg NOISE)"/>
        <param name="du_0"        type="double" value="0.01" if="$(arg NOISE)"/>
        <param name="du_1"        type="double" value="0.05" if="$(arg NOISE)"/>
        <param name="u_bound"     type="double" value="1.0" if="$(arg NOISE)"/>


        <param name="n_bound"     type="double" value="0.0" unless="$(arg NOISE)"/>
        <param name="x_std"       type="double" value="0.0" unless="$(arg NOISE)"/>
        <param name="y_std"       type="double" value="0.0" unless="$(arg NOISE)"/>
        <param name="yaw_std"     type="double" value="0.0" unless="$(arg NOISE)"/>
        <param name="vx_std"       type="double" value="0.0" unless="$(arg NOISE)"/>
        <param name="vy_std"       type="double" value="0.0" unless="$(arg NOISE)"/>
        <param name="omega_std"  type="double" value="0.0" unless="$(arg NOISE)"/>
        <param name="du_0"        type="double" value="0.0" unless="$(arg NOISE)"/>
        <param name="du_1"        type="double" value="0.0" unless="$(arg NOISE)"/>
        <param name="u_bound"     type="double" value="0.0" unless="$(arg NOISE)"/>


        <!-- drag force parameters -->
        
        <!--  Initialize the states -->
        <?ignore
        <param name="init_x"    type="double" value="1.16713867" />
        <param name="init_y"    type="double" value="0.34539007" />
        <param name="init_yaw"  type="double" value="0" />
        ?>

        <param name="init_x"    type="double" value="0.1" />
        <param name="init_y"    type="double" value="0.0" />
        <param name="init_yaw"  type="double" value="0.0" />

    </node>

    <!--node pkg="simulator" type="2Dplot.py" name="visualization" output="screen">
    </node>-->

    <?ignore
    <!-- TRAJECTORY PLANNER -->
    <node pkg="lpv_mpc" type="plannerMain.py" name="TrajectoryPlanner" output="screen">
        <!--No change in planning track if value is changed-->
        <param name="Planning_Track"   type="int"      value="3" />
        <param name="max_vel"          type="double"   value="5.0" />
        <param name="min_vel"          type="double"   value="0.0" />
        <param name="N"                type="int"      value="20" />
        <param name="Frecuency"        type="int"      value="20" />
        <param name="halfWidth"        type="double"   value="0.8" />
        <param name="Visualization"    type="int"      value="0" />
        <param name="Testing"          type="int"      value="0" />
    </node>


    <!--DYNAMIC OBSERVER (STATE ESTIMATOR)-->
    <node pkg="lpv_mpc"              type="stateEstimator.py" name="state_estimator" output="screen">
        <!-- delay in state estimator -->
         <param name="delay_df"   type="double" value="0.0" />
         <param name="delay_a"    type="double" value="0.0" />
         <param name="estimatorON" type="bool"  value="true"/>
         <!-- ekf properties -->
          <param name="thReset"   type="double" value="0.5" />
          <param name="vSwitch"   type="double" value="1.3" />
          <param name="psiSwitch" type="double" value="1.2" />
          <param name="loopRate"  type="double" value="200" />
    </node>

    <!-- CONTROL -->
    <node pkg="lpv_mpc" type="controllerMain.py" name="control" output="log">
        <param name="mode"        type="str"     value="simulations" />
        <param name="saveData"    type="bool"    value="true" />
        <param name="N"           type="int"     value="20" />
        <param name="Hz"          type="double"  value="30" />
        <param name="vel_ref"     type="double"  value="2.5"/>
        <param name="vel_th"      type="double"  value="0.4" />
    </node>
    ?>

</launch>
